provider/core/src/resource.rs

```diff
@@ -74,59 +73,132 @@ impl writeable::Writeable for ResourceCategory {
+/// Executables can be searched for ResourceKey instances to produce optimized data files.
+/// Therefore, users should not generally create ResourceKey instances; they should instead use
+/// the ones exported by a component.
+#[derive(PartialEq, Eq, Copy, Clone)]
+#[repr(C)]
 pub struct ResourceKey {
-    pub category: ResourceCategory,
-    pub sub_category: TinyStr16,
-    pub version: u16,
+    path: &'static str,
+    _tag0: [u8; 8],
+    hash: ResourceKeyHash,
+    _tag1: [u8; 2],
 }
 
-/// Shortcut to construct a const resource identifier.
-///
-/// # Examples
-///
-/// Create a private-use ResourceKey:
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(x, "foo", "bar", 1);
-/// assert_eq!("x-foo/bar@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-///
-/// Create a ResourceKey for a specific [`ResourceCategory`] (for ICU4X library code only):
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(Plurals, "ordinal", 1);
-/// assert_eq!("plurals/ordinal@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-#[macro_export]
-macro_rules! resource_key {
-    ($category:ident, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!($crate::ResourceCategory::$category, $sub_category, $version)
-    };
-    (x, $pu:literal, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!(
-            $crate::ResourceCategory::PrivateUse($crate::internal::tinystr4!($pu)),
-            $sub_category,
-            $version
-        )
-    };
-    ($category:expr, $sub_category:literal, $version:tt) => {
-        $crate::ResourceKey {
-            category: $category,
-            sub_category: $crate::internal::tinystr16!($sub_category),
-            version: $version,
+impl ResourceKey {
+    /// Gets a human-readable representation of a [`ResourceKey`].
+    /// 
+    /// The human-readable path string always contains at least one '/', and it ends with '@'
+    /// followed by one or more digits. Paths do not contain characters other than lowercase
+    /// ASCII, '_', '/', and '@'.
+    /// 
+    /// Useful for reading and writing data to a file system.
+    #[inline]
+    pub const fn get_path(&self) -> &str {
+        self.path
+    }
+
+    /// Gets a machine-readable representation of a [`ResourceKey`].
+    ///
+    /// The machine-readable hash is 4 bytes and can be used as the key in a map.
+    ///
+    /// The hash is a 32-bit FxHash of the path, computed as if on a little-endian platform.
+    #[inline]
+    pub const fn get_hash(&self) -> ResourceKeyHash {
+        self.hash
+    }
+
+    /// Creates a new ResourceKey from a path, panicking if the path is invalid.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the syntax of the path is not valid.
+    #[inline]
+    pub const fn new(path: &'static str) -> Self {
```

Comment:
note that the rust compiler will happily generate (potentially expensive) runtime code for this when this is called in a non const context: i recommend using a macro the way my tinystr PR does to force it to be a static. That way you can also guarantee that the hash will be in the source and not generated at runtime.

Reply:
I don't know if I agree.

1. We cannot prevent people from using the function, because the macro is just codegen and it needs to wrap a public function (although we could rename or hide the function).
2. The type name is included in the prelude and it's more clear what the function is for (it creates a new ResourceKey, by definition); macros are relatively less friendly for documentation and tooling.
3. My next step is going to be adding a `const ResourceKey` field to `DataMarker`, which will force these into a const context.
4. The problem of const functions "accidentally" resolving at runtime instead of build time is not unique to us. I'd rather not introduce the concept of using macros to solve this problem unless you can point to something that says that this is the "official" way to solve the problem.

Reply:
In this case because it's going to be rare to construct these on your own I'm less worried about _preventing_ people from using this, and more about giving people a convenient way to instantiate a const. There is no such convenient way currently.

But given that most users will just use the exported const, I guess it's fine.

I don't have anything saying this is the "official" way to solve the footgun but it's basically the only way and I asked other community members: the const thing happens because rust doesn't really pick whether to const a function based on whether the argument is a literal, and the only way to say "my arguments must be literals" is via a macro. It's the right tool for the job.

Reply:
Yes, the typical way to do that is to remove all constructors (except often `Default::default()`), leave in a `#[doc(hidden)]` constructor with a name that has underscores in it (`ResourceKey::__internal_construct()` or something), and use it from the macro.

I don't think we _need_ to here though

Reply:
I added back the macro, and kept the `try_new` constructor but not the `new` constructor, with appropriate documentation.